internal/utils/csi.go

```diff
@@ -0,0 +1,393 @@
+			FieldPath: "metadata.namespace",
+		},
+	},
+}
+var PodUidEnvVar = corev1.EnvVar{
+	Name: "POD_UID",
+	ValueFrom: &corev1.EnvVarSource{
+		FieldRef: &corev1.ObjectFieldSelector{
+			FieldPath: "metadata.uid",
+		},
+	},
+}
+var NodeIdEnvVar = corev1.EnvVar{
+	Name: "NODE_ID",
+	ValueFrom: &corev1.EnvVarSource{
+		FieldRef: &corev1.ObjectFieldSelector{
+			FieldPath: "spec.nodeName",
+		},
+	},
+}
+var DriverNamespaceEnvVar = corev1.EnvVar{
+	Name: "DRIVER_NAMESPACE",
+	ValueFrom: &corev1.EnvVarSource{
+		FieldRef: &corev1.ObjectFieldSelector{
+			FieldPath: "metadata.namespace",
+		},
+	},
+}
+
+// CSI Addons container port definition
+var CsiAddonsContainerPort = corev1.ContainerPort{
+	ContainerPort: int32(9070),
+}
+
+// Ceph CSI common container arguments
+var CsiAddressContainerArg = fmt.Sprintf("--csi-address=%s", csiEndpoint)
+var EndpointContainerArg = fmt.Sprintf("--endpoint=%s", csiEndpoint)
+var CsiAddonsEndpointContainerArg = fmt.Sprintf("--csi-addons-endpoint=(%s)", csiAddonsEndpoint)
+var CsiAddonsAddressContainerArg = fmt.Sprintf("--csi-addons-address=(%s)", csiAddonsEndpoint)
+var LeaderElectionContainerArg = "--leader-election=true"
+var NodeIdContainerArg = fmt.Sprintf("--nodeid=$(%s)", NodeIdEnvVar.Name)
+var PidlimitContainerArg = "--pidlimit=-1"
+var ControllerServerContainerArg = "--controllerserver=true"
+var NodeServerContainerArg = "--nodeserver=true"
+var RetryIntervalStartContainerArg = "--retry-interval-start=500ms"
+var DefaultFsTypeContainerArg = "--default-fstype=ext4"
+var HandleVolumeInuseErrorContainerArg = "--handle-volume-inuse-error=false"
+var PodUidContainerArg = fmt.Sprintf("--pod-uid=$(%s)", PodUidEnvVar.Name)
+var PodContainerArg = fmt.Sprintf("--pod=$(%s)", PodNameEnvVar.Name)
+var NamespaceContainerArg = fmt.Sprintf("--namespace=(%s)", PodNamespaceEnvVar.Name)
+var ControllerPortContainerArg = fmt.Sprintf("--controller-port=%d", CsiAddonsContainerPort.ContainerPort)
+var DriverNamespaceContainerArg = fmt.Sprintf("--drivernamespace=$(%s)", DriverNamespaceEnvVar.Name)
+var MetricsPathContainerArg = "--metricspath=/metrics"
+var PoolTimeContainerArg = "--polltime=60s"
+var ExtraCreateMetadataContainerArg = "--extra-create-metadata=true"
+var PreventVolumeModeConversionContainerArg = "--prevent-volume-mode-conversion=true"
+var HonorPVReclaimPolicyContainerArg = "--feature-gates=HonorPVReclaimPolicy=true"
+var TopologyContainerArg = "--feature-gates=Topology=true"
+var RecoverVolumeExpansionFailureContainerArg = "--feature-gates=RecoverVolumeExpansionFailure=true"
+var EnableVolumeGroupSnapshotsContainerArg = "--enable-volume-group-snapshots=true"
+var ForceCephKernelClientContainerArg = "--forcecephkernelclient=true"
+
+func LogLevelContainerArg(level int) string {
+	return fmt.Sprintf("--v=%d", Clamp(level, 0, 5))
+}
+func TypeContainerArg(t string) string {
+	switch t {
+	case "rbd", "cephfs", "nfs", "controller", "liveness":
+		return fmt.Sprintf("--type=%s", t)
+	default:
+		return ""
```

Comment:
the caller should handle this as error or panic if anything else is passed

Reply:
Do you mean that we want to have a reconcile error if this arge is not set? The same mentality can be applied to many other flags.

I think these class of concerns are less of an operator failure and more as operand failure which we will report in the status. Either way re-queuing will not help here

Reply:
okay what is the expected result if we get into the default state here?

Reply:
I would expect that the sidecar will not be healthy and that we will report it on our status section (we still need to have a discussion about the status section). 
I assume that we can get meaningful information from a sidecar failure that we can use as the reason